src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,245 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright contributors to the Ceph project
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "topic_migration.h"
+#include "services/svc_zone.h"
+#include "rgw_sal_rados.h"
+
+namespace rgwrados::topic_migration {
+
+namespace {
+
+// migrate v1 notification metadata for a single bucket
+int migrate_notification(const DoutPrefixProvider* dpp, optional_yield y,
+                         rgw::sal::RadosStore* driver, const rgw_raw_obj& obj)
+{
+  // parse bucket name and marker of out "pubsub.{tenant}.bucket.{name}/{marker}"
+  std::string tenant;
+  std::string bucket_name;
+  std::string marker;
+  int r = deconstruct_topics_oid(obj.oid, tenant, bucket_name, marker, dpp);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read tenant, bucket name and marker from " << obj
+                      << " with: " << cpp_strerror(r)
+                      << " expected format pubsub.{tenant}.bucket.{name}/{marker}!" << dendl;
+    return r;
+  }
+
+  // migrate the notifications
+  rgw_pubsub_bucket_topics bucket_topics;
+  rgw_bucket rgw_bucket_info(tenant, bucket_name);
+  rgw_bucket_info.marker = marker;
+  rgw::sal::RadosBucket rados_bucket(driver, rgw_bucket_info);
+  RGWObjVersionTracker bucket_topics_objv;
+  r = rados_bucket.read_topics(bucket_topics, &bucket_topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read v1 bucket notifications from " << obj
+                      << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // in a for-loop that retries ECANCELED errors:
+  // {
+  // load the corresponding bucket by name
+  // break if marker doesn't match loaded bucket's
+  // break if RGW_ATTR_BUCKET_NOTIFICATION xattr already exists
+  // write RGW_ATTR_BUCKET_NOTIFICATION xattr
+  // }
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  do {
+    r = driver->load_bucket(dpp, rgw_bucket_info, &bucket, y);
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "failed to load the bucket from " << obj
+                        << " with: " << cpp_strerror(r) << dendl;
+      continue;
```

Comment:
if the bucket doesn't exist, `continue` here would end up in an infinite loop

if `r == -ENOENT`, the bucket is gone and we should delete the v1 notification

any other error should be logged and returned without deleting the v1 notification object. it would need to be retried later

Reply:
Ack!